comp/networkpath/npscheduler/npschedulerimpl/pathteststore/pathteststore.go

```diff
@@ -0,0 +1,144 @@
+	// are called by different routines.
+	pathtestConfigsMutex sync.Mutex
+
+	// pathtestInterval defines how frequently pathtests should run
+	pathtestInterval time.Duration
+
+	// pathtestTTL is the duration a Pathtest should run from discovery.
+	// If a Pathtest is added again before the TTL expires, the TTL is reset to this duration.
+	pathtestTTL time.Duration
+}
+
+func newPathtestContext(pt *common.Pathtest, runUntilDuration time.Duration) *PathtestContext {
+	now := timeNow()
+	return &PathtestContext{
+		Pathtest:     pt,
+		nextRunTime:  now,
+		runUntilTime: now.Add(runUntilDuration),
+	}
+}
+
+// NewPathtestStore creates a new PathtestStore
+func NewPathtestStore(pathtestTTL time.Duration, pathtestInterval time.Duration, logger log.Component) *PathtestStore {
+	return &PathtestStore{
+		pathtestContexts: make(map[uint64]*PathtestContext),
+		pathtestTTL:      pathtestTTL,
+		pathtestInterval: pathtestInterval,
+		logger:           logger,
+	}
+}
+
+// Flush will flush specific Pathtest context (distinct hash) if nextRunTime is reached
+// once a Pathtest context is flushed nextRunTime will be updated to the next flush time
+//
+// pathtestTTL:
+// pathtestTTL defines the duration we should keep a specific PathtestContext in `PathtestStore.pathtestContexts`
+// after `lastSuccessfulFlush`. // Flow context in `PathtestStore.pathtestContexts` map will be deleted if `pathtestTTL`
+// is reached to avoid keeping Pathtest context that are not seen anymore.
+// We need to keep PathtestContext (contains `nextRunTime` and `lastSuccessfulFlush`) after flush
+// to be able to flush at regular interval (`flushInterval`).
+// Example, after a flush, PathtestContext will have a new nextRunTime, that will be the next flush time for new pathtestContexts being added.
+func (f *PathtestStore) Flush() []*PathtestContext {
+	f.pathtestConfigsMutex.Lock()
+	defer f.pathtestConfigsMutex.Unlock()
+
+	f.logger.Tracef("f.pathtestContexts: %+v", f.pathtestContexts)
+	// DEBUG STATEMENTS
+	for _, ptConf := range f.pathtestContexts {
+		if ptConf.Pathtest != nil {
+			f.logger.Tracef("in-mem ptConf %s:%d", ptConf.Pathtest.Hostname, ptConf.Pathtest.Port)
+		}
+	}
+
+	var pathtestsToFlush []*PathtestContext
+	for key, ptConfigCtx := range f.pathtestContexts {
+		now := timeNow()
+
+		if ptConfigCtx.runUntilTime.Before(now) {
+			f.logger.Tracef("Delete Pathtest context (key=%d, runUntilTime=%s, nextRunTime=%s)", key, ptConfigCtx.runUntilTime.String(), ptConfigCtx.nextRunTime.String())
+			// delete ptConfigCtx wrapper if it reaches runUntilTime
+			delete(f.pathtestContexts, key)
+			continue
+		}
+		if ptConfigCtx.nextRunTime.After(now) {
+			continue
+		}
+		if !ptConfigCtx.lastFlushTime.IsZero() {
+			ptConfigCtx.lastFlushInterval = now.Sub(ptConfigCtx.lastFlushTime)
+		}
+		ptConfigCtx.lastFlushTime = now
+		pathtestsToFlush = append(pathtestsToFlush, ptConfigCtx)
+		ptConfigCtx.nextRunTime = ptConfigCtx.nextRunTime.Add(f.pathtestInterval)
```

Comment:
Is it ok if `nextRunTime > runUntilTime` after this?

Reply:
Would this guarantee that `nextRunTime > now`? If not, would that mean that the context gets flushed more than once?

Reply:
So in the second case, we may run two contexts back to back, instead of an interval?

Reply:
If we are already backed up, wouldn't this strategy make things worse?

Reply:
Right now, I would like to keep the current implemented strategy if that's fine for you. I mean this one:

> It doesn't guarantee that, but I think it's not an issue. If nextRunTime > now, it usually mean that we are behind schedule (late) for running some of the traceroute. And indeed, contexts might be flushed multiple time at each flush if it's very late behind schedule, but that's fine since we are just "caching up".